### PR TITLE
Vital update on json.md about nested classes

### PR DESCRIPTION
**Pull request by marcelino-borges**
A super relevant subject is missing at this documentation and is so dificult to find in the web: how to set the program to generate the json properly when we have nested classes in our models. I have spent 2 days trying to figure it out when I could have killed this at the first moment when I first read this doc trying to learn how to use the json generation...

### CODE REVIEW

**Comment by marcelino-borges** [comment]
@googlebot I signed it!

File: src/docs/development/data-and-backend/json.md [hunk 1]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+  String city;
+  
+  Address(this.street, this.city);
+  
+  factory Address.fromJson(Map<String, dynamic> json) => _$AddressFromJson(json);
+  Map<String, dynamic> toJson() => _$AddressToJson(this); 
+}
+```
+
+And the class Address is used (nested) by this our User class:
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: Replace with:

The `Address` class is nested inside the `User` class:

File: src/docs/development/data-and-backend/json.md [hunk 2]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+  Address address;
+  
+  User(this.firstName, this.address);
+
+  factory User.fromJson(Map<String, dynamic> json) => _$UserFromJson(json);
+  Map<String, dynamic> toJson() => _$UserToJson(this);
+}
+```
+
+If you run `flutter pub run build_runner build` in the terminal, the `*.g.dart` will be created but the private _$UserToJson() method created inside this new file will be more or less like this:
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: Replace with:
```
Running `flutter pub run build_runner build` in the terminal creates
the `*.g.dart` file, but the private `_$UserToJson()` function
looks something like the following:
```

File: src/docs/development/data-and-backend/json.md [hunk 3]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+
+All looks fine right now, but if you do a print() on our user object:
+
+```dart
+Address address = Address("My st.", "New York");
+User user = User("John", address);
+print(user.toJson());
+```
+
+The result will be:
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: => The result is:  

(Avoid future tense. Rule: https://developers.google.com/style/tense )

File: src/docs/development/data-and-backend/json.md [hunk 4]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+print(user.toJson());
+```
+
+The result will be: 
+
+```json
+{name: John, adress: Instance of 'address'}
+```
+
+And that's not what we want. We need this address being parsed to json as well as all the rest, to have something like:
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: Replace with:

```
When what you probably want is output like the following:
```

File: src/docs/development/data-and-backend/json.md [hunk 5]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+
+```json
+{name: John, adress: {street: My st., city: New York}}
+```
+
+By default, there is a parameter called `explicitToJson` set to false which doesn't allow our code to get a `.toJson()` after our `instance.address` passed as the value of the address key, in our returning json from `_$UserToJson()`.
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: I think this is unnecessarily detailed and implementation specific. I'd delete this paragraph.

File: src/docs/development/data-and-backend/json.md [hunk 6]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+
+In order to make this work we have to pass this `explicitToJson` parameter as `true` in our `@JsonSerializable()` annotations. So our User's class stays like this now:
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: Replace with:
```
To make this work, pass `explicitToJson: true` in the `@JsonSerializable()`
annotation. The `User` class now looks as follows:
```

File: src/docs/development/data-and-backend/json.md [hunk 7]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+  String firstName;  
+  Address address;
+  
+  User(this.firstName, this.address);
+
+  factory User.fromJson(Map<String, dynamic> json) => _$UserFromJson(json);
+  Map<String, dynamic> toJson() => _$UserToJson(this);
+}
+```
+
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: I'd delete everything from here....

File: src/docs/development/data-and-backend/json.md [hunk 8]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+The result will be: 
+
+```json
+{name: John, adress: Instance of 'address'}
+```
+
+And that's not what we want. We need this address being parsed to json as well as all the rest, to have something like:
+
+```json
+{name: John, adress: {street: My st., city: New York}}
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: adress -> address

File: src/docs/development/data-and-backend/json.md [hunk 9]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+
+The same would be applied if we had a `List<Address> addresses` in our User class. This list would only be parsed to json if our explicitToJson is set to true, so we would have a `_$UserToJson()` like this: 
+
+``` dart
+(
+Map<String, dynamic> _$UserToJson(User instance) => <String, dynamic>{      
+  'firstName': instance.firstName,
+  'addresses': instance.addresses?.map((e) => e?.toJson())?.toList(),      
+};
+```
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: ...to here. It's not needed.

File: src/docs/development/data-and-backend/json.md [hunk 10]
```diff
@@ -364,6 +361,120 @@ to write automated tests to ensure that the serialization works&mdash;it's
...
+
+For more info: [JsonSerializable class](https://pub.dev/documentation/json_annotation/latest/json_annotation/JsonSerializable-class.html) and [explicitToJson](https://pub.dev/documentation/json_annotation/latest/json_annotation/JsonSerializable/explicitToJson.html?fbclid=IwAR1Y_KPvOVQw8gDnxps68YbLgr2UdddePXkpJgcCe5edf5fOSYTffBelRw8)
```
Comment (sfshaza2) [inline src/docs/development/data-and-backend/json.md]: Replace with:

```
For more information, see [explicitToJson][] in the 
[JsonSerializable][] class for the [json_annotation][] package.

[explicitToJson]: {{site.pub}}/documentation/json_annotation/latest/json_annotation/JsonSerializable/explicitToJson.html
[JsonSerializable]: {{site.pub}}/documentation/json_annotation/latest/json_annotation/JsonSerializable-class.html
[json_annotation]: {{site.pub}}/packages/json_annotation
```

**Comment by marcelino-borges** [comment]
Thank you @sfshaza2 ! Sorry for the mistakes and broken rules. It's my first contribution here. I hope this can someone. I suffered a bit trying to figure this out. I guess that having this in the docs can really save some time of work...

**Comment by sfshaza2** [comment]
@marcelino-borges, please don't apologize! This is a wonderful contribution to our docs! We appreciate you taking the time to share your hard-won knowledge! Just let me know if you are up to making the edits.

**Comment by marcelino-borges** [comment]
@sfshaza2 , I have already commited the changes you mentioned.

**Comment by marcelino-borges** [comment]
@sfshaza2 Can you tell me how is the proccess until the merge? Or could you please tell me where I can read about it? Sorry for being such a noob... ^^'

**Pull request merged**